src/tinyuf2_board.c

```diff
@@ -13,31 +13,54 @@
 #include "ice_flash.h"
 #include "ice_fpga.h"
 #include "ice_spi.h"
+#include "ice_sram.h"
 
 static alarm_id_t alarm_id;
-static bool flash_ready;
+static bool spi_ready;
+static uint32_t flash_erased[ICE_FLASH_SIZE_BYTES / (ICE_FLASH_BLOCK_SIZE * 32)];
+
+#define SRAM_ADDR 0x20000000
 
 void board_flash_read(uint32_t addr, void *buffer, uint32_t len)
 {
-    if (!flash_ready) {
+    if (!spi_ready) {
+        ice_fpga_stop();
+        ice_sram_init();
         ice_flash_init();
-        flash_ready = true;
+        spi_ready = true;
+    }
+    if (addr >= SRAM_ADDR) {
+        addr -= SRAM_ADDR;
+        ice_sram_read_blocking(addr, buffer, len);
+    }
+    else {
+        ice_flash_read(addr, buffer, len);
     }
-    ice_flash_read(addr, buffer, len);
 }
 
 void board_flash_write(uint32_t addr, const void *data, uint32_t len)
 {
-    if (!flash_ready) {
+    if (!spi_ready) {
         ice_fpga_stop();
+        ice_sram_init();
         ice_flash_init();
-        flash_ready = true;
-        ice_flash_erase_chip();
+        spi_ready = true;
+    }
+    if (addr >= SRAM_ADDR) {
+        addr -= SRAM_ADDR;
+        ice_sram_write_blocking(addr, data, len);
     }
-    if (len != ICE_FLASH_PAGE_SIZE) {
-        printf("%s: expected len=%u got len=%ld\r\n", __func__, ICE_FLASH_PAGE_SIZE, len);
-    } else {
-        ice_flash_program_page(addr, data);
+    else {
+        int flash_erase_idx = addr / ICE_FLASH_BLOCK_SIZE;
+        if ((flash_erased[flash_erase_idx >> 5] & (1u << (flash_erase_idx & 0x1F))) == 0) {
+            ice_flash_erase_block(addr & ~(ICE_FLASH_BLOCK_SIZE - 1));
+            flash_erased[flash_erase_idx >> 5] |= 1u << (flash_erase_idx & 0x1F);
+        }
+        if (len != ICE_FLASH_PAGE_SIZE) {
+            printf("%s: expected len=%u got len=%ld\r\n", __func__, ICE_FLASH_PAGE_SIZE, len);
+        } else {
+            ice_flash_program_page(addr, data);
+        }
```

Comment:
I think I got lost: why using only 5 bits out of the 8 of a byte for the bitmap?
I mean `& 0x1F` vs `& 0xFF` along with `flash_erase_idx >> 8` instead of `flash_erase_idx >> 5`.

Reply:
Maybe to have some kind of granularity that fits char arrays very well?
To avoid computing slices of bits that are split between two bytes?

Reply:
This is a bitmap stored in an array of uint32s.  

The `& 0x1F` gives a number between 0 and 31 corresponding to the bit within the array entry, and we then check/set that bit.

The `>> 5` shift out the low 5 bits (that were masked when selecting the bit), in order to select the array entry.

Reply:
Not sure why, I thought this was an array of uint8_t... This makes a lot more sense now. Thank you!
The current implementation is ideal then.
I learned plenty!